xmr-btc/src/lib.rs

```diff
@@ -50,3 +50,151 @@ pub mod bitcoin;
 pub mod bob;
 pub mod monero;
 pub mod transport;
+
+use ecdsa_fun::{adaptor::Adaptor, nonce::Deterministic};
+use genawaiter::sync::{Gen, GenBoxed};
+use sha2::Sha256;
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug)]
+pub enum Action {
+    LockBitcoin(bitcoin::TxLock),
+    SendBitcoinRedeemEncsig(bitcoin::EncryptedSignature),
+    CreateMoneroWalletForOutput {
+        spend_key: monero::PrivateKey,
+        view_key: monero::PrivateViewKey,
+    },
+    RefundBitcoin {
+        tx_cancel: bitcoin::Transaction,
+        tx_refund: bitcoin::Transaction,
+    },
+}
+
+// TODO: This could be moved to the monero module
+pub trait ReceiveTransferProof {
+    fn receive_transfer_proof(&self) -> monero::TransferProof;
+}
+
+/// Perform the on-chain protocol to swap monero and bitcoin as Bob.
+///
+/// This is called post handshake, after all the keys, addresses and most of the
+/// signatures have been exchanged.
+pub fn action_generator_bob<N, M, B>(
+    network: &'static N,
+    monero_ledger: &'static M,
+    bitcoin_ledger: &'static B,
+    // TODO: Replace this with a new, slimmer struct?
+    bob::State2 {
```

Comment:
Why only start at `State2`? Could pack the whole protocol into here :)

Reply:
It's certainly an option, but @tcharding is working on the libp2p protocol at the moment. I thought that it would be easier (at least for now), to separate the initial handshake (during which signatures and keys are exchanged) from the on-chain protocol, so that we can work in parallel.

Reply:
I guess it depends on how you design the libp2p protocol. Are they always the same messages, unconditionally? If so then it could probably be encapsulated into a single protocol at the `ProtocolsHandler` level and only emit the completed handshake to the `NetworkBehaviour` and sub-sequentially to the rest of the application.

If that is the case, then it would probably also split the protocol logically into these parts so that what you are doing here requires the handshake to be done. Currently it feels like we are starting "in the middle". Maybe that is fine though :)